### ISSUE TITLE
Question on FP16/BF16 loss accumulation over large token counts

### ISSUE THREAD

**Issue by francesco-bertolotti** [pytorch/torchtitan#2336]
I noticed that the loss computation was changed from averaging to summing and then dividing by the number of valid tokens ([[commit](https://github.com/pytorch/torchtitan/commit/0cb743558a7aa96eabbc099d1712f85b197e1edb)](https://github.com/pytorch/torchtitan/commit/0cb743558a7aa96eabbc099d1712f85b197e1edb)).

I’m wondering whether performing this accumulation in `float16`/`bfloat16` could introduce compounded numerical error, potentially leading to instability when summing losses over very large token counts (e.g., 1–2M tokens).

Do you have any insights on this? Perhaps this has already been evaluated and found to be stable in practice.

@wwwjn @fegin

**Comment by wwwjn** [comment]
Thanks for raising this, I think we should have constant and clear information about the data type. Can you explain why you think accumulating is in float16, or when will the accumulation in float16? The loss function called .float() before it returns, and the `global_valid_token` is an int. So the loss is accumulated in float32.

Or are you saying you want to perform accumulation in float16, which is different from torchtitan's behavior?

**Comment by francesco-bertolotti** [comment]
My mistake. I overlooked the float32 cast. I ran a few tests, and the differences are indeed negligible when calling prediction.float() before computing the loss. Thanks for pointing this out.
